src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -0,0 +1,231 @@
+package is.hail.io
+
+import is.hail.HailContext
+import is.hail.annotations._
+import is.hail.expr._
+import is.hail.sparkextras.OrderedRDD2
+import is.hail.utils._
+import is.hail.variant._
+import org.apache.hadoop.conf.Configuration
+import org.apache.spark.rdd.RDD
+
+import scala.language.implicitConversions
+import scala.io.Source
+
+object LoadMatrix {
+
+  def warnDuplicates(ids: Array[String]) {
+    val duplicates = ids.counter().filter(_._2 > 1)
+    if (duplicates.nonEmpty) {
+      warn(s"Found ${duplicates.size} duplicate ${plural(duplicates.size, "sample ID")}:\n  @1",
+        duplicates.toArray.sortBy(-_._2).map { case (id, count) => s"""($count) "$id"""" }.truncatable("\n  "))
+    }
+  }
+
+  // this assumes that col IDs are in last line of header.
+  /// FIXME: Is the toString.split call too slow?
+  def parseHeader(lines: Array[String], sep: String = "\t"): Array[String] = {
+    lines.last.toString.split(sep)
+  }
+
+  def getHeaderLines[T](hConf: Configuration, file: String, nLines: Int = 1): Array[String] = hConf.readFile(file) { s =>
+    Source.fromInputStream(s)
+      .getLines()
+      .take(nLines)
+      .toArray
+  }
+
+  def apply(hc: HailContext,
+            file1: String,
+            files: Array[String],
+            nPartitions: Option[Int] = None,
+            dropSamples: Boolean = false,
+            sep: String = "\t"):
+  VariantSampleMatrix[Annotation, Annotation, Annotation] = {
+    val sc = hc.sc
+    val hConf = hc.hadoopConf
+
+    val header1 = parseHeader(getHeaderLines(hConf, file1), sep)
+    val header1Bc = sc.broadcast(header1)
+
+    val sampleIds: Array[String] =
+      if (dropSamples)
+        Array.empty
+      else
+        header1
+
+    val nSamples = sampleIds.length
+
+    LoadMatrix.warnDuplicates(sampleIds)
+
+    val lines = sc.textFilesLines(files, nPartitions.getOrElse(sc.defaultMinPartitions))
+
+    val fileByPartition = lines.partitions.map(p => partitionPath(p))
+    val firstPartitions = fileByPartition.zipWithIndex.filter((name) => name._2 == 0 || fileByPartition(name._2-1) != name._1).map((name) => name._2)
+
+    val matrixType = MatrixType(VSMMetadata(
+      sSignature = TString,
+      vSignature = TString,
+      genotypeSignature = TInt64
+    ))
+
+    val keyType = matrixType.kType
+    val rowKeys: RDD[RegionValue] = lines.mapPartitionsWithIndex { (i,it) =>
+
+      if (firstPartitions.contains(i)) {
+        val hd1 = header1Bc.value
+        val hd = it.next().value.split(sep)
+        if (!hd1.sameElements(hd)) {
+          hd1.zipAll(hd, None, None).zipWithIndex.foreach { case ((s1, s2), j) =>
```

Comment:
Hmm, not sure about this.

There's some type trickiness going.  `s1`, and `s2` have type `Any`, because that's the common supertype of String (type element type of `hd1` and `hd`) and `None`, which has type `Option[Nothing]`.

Second, it will print in the error message as `None`, but you don't distinguish between the sample ID being missing and the sample ID having the value "None".  In particular, if you had a file with one extra sample ID "None", you'd get a message like: found different sample ids at position 10:
  file1: None
  file2: None